protocol/daemons/server/types/health_monitor.go

```diff
@@ -0,0 +1,174 @@
+package types
+
+import (
+	cosmoslog "cosmossdk.io/log"
+	"fmt"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/dydxprotocol/v4-chain/protocol/daemons/types"
+	libtime "github.com/dydxprotocol/v4-chain/protocol/lib/time"
+	"sync"
+	"time"
+)
+
+const (
+	// HealthCheckPollFrequency is the frequency at which the health checkable service is polled.
+	HealthCheckPollFrequency = 5 * time.Second
+)
+
+// HealthMonitor monitors the health of daemon services, which implement the HealthCheckable interface. If a
+// registered health-checkable service sustains an unhealthy state for the maximum acceptable unhealthy duration,
+// the monitor will execute a callback function.
+type HealthMonitor struct {
+	// serviceToHealthChecker maps daemon service names to their update metadata.
+	serviceToHealthChecker map[string]*healthChecker
+	// stopped indicates whether the monitor has been stopped. Additional daemon services cannot be registered
+	// after the monitor has been stopped.
+	stopped bool
+	// disabled indicates whether the monitor has been disabled. This is used to disable the monitor in testApp
+	// tests, where app.New is not executed.
+	disabled bool
+	// lock is used to synchronize access to the monitor.
+	lock sync.Mutex
```

Comment:
I walked through the code and confirmed that, according to my understanding, health monitor or health checker locks are being called in a nested fashion. I also reduced the critical section for the health monitor Stop method just in case.